agent/wsclient/client_test.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2014-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package wsclient
+
+import (
+	"io"
+	"os"
+	"testing"
+
+	"github.com/aws/amazon-ecs-agent/agent/acs/model/ecsacs"
+	"github.com/aws/amazon-ecs-agent/agent/config"
+	"github.com/aws/amazon-ecs-agent/agent/wsclient/mock/utils"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+
+	"github.com/gorilla/websocket"
+
+	"github.com/stretchr/testify/assert"
+)
+
+const dockerEndpoint = "/var/run/docker.sock"
+
+// TestConcurrentWritesDontPanic will force a panic in the websocket library if
+// the implemented methods don't handle concurrency correctly
+// See https://godoc.org/github.com/gorilla/websocket#hdr-Concurrency
+func TestConcurrentWritesDontPanic(t *testing.T) {
+	mockServer, _, requests, _, _ := mockwsutils.StartMockServer(t, make(chan []byte))
+	defer mockServer.Close()
+
+	req := ecsacs.AckRequest{Cluster: aws.String("test"), ContainerInstance: aws.String("test"), MessageId: aws.String("test")}
+
+	cs := getClientServer(mockServer.URL)
+	cs.Connect()
+
+	executeTenRequests := func() {
+		for i := 0; i < 10; i++ {
+			cs.MakeRequest(&req)
+		}
+	}
+
+	// Make requests from two separate routines to try and force a
+	// concurrent write
+	go executeTenRequests()
+	executeTenRequests()
+
+	t.Log("Waiting for all 20 requests to succeed")
+	for i := 0; i < 20; i++ {
+		<-requests
+	}
+}
+
+// TestProxyVariableCustomValue ensures that a user is able to override the
+// proxy variable by setting an environment variable
+func TestProxyVariableCustomValue(t *testing.T) {
+	mockServer, _, _, _, _ := mockwsutils.StartMockServer(t, make(chan []byte))
+	defer mockServer.Close()
+
+	testString := "Custom no proxy string"
+	os.Setenv("NO_PROXY", testString)
+	getClientServer(mockServer.URL).Connect()
+
+	assert.Equal(t, os.Getenv("NO_PROXY"), testString, "NO_PROXY should match user-supplied variable")
+}
+
+// TestProxyVariableDefaultValue verifies that NO_PROXY gets overridden if it
+// isn't already set
+func TestProxyVariableDefaultValue(t *testing.T) {
+	mockServer, _, _, _, _ := mockwsutils.StartMockServer(t, make(chan []byte))
```

Comment:
Will they get torn down at the end of each test? I didn't think it would persist.
If closing them is better I can do that too.